include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,396 @@
+        }));
+
+        current_lower_bound = current_upper_bound;
+        current_upper_bound += interval;
+    }
+
+    prime_vectors[ranges].reserve(primes_in_range);
+    future_manager.emplace_back(std::async(std::launch::async, [current_lower_bound, upper_bound, &primes, &prime_vectors]{
+        boost::math::detail::IntervalSieve sieve(current_lower_bound, upper_bound, primes, prime_vectors.back());
+    }));
+
+    for(auto &&future : future_manager)
+    {
+        if(future.valid())
+        {
+            future.get();
+        }
+    }
+
+    for(auto &v : prime_vectors)
+    {
+        resultant_primes.insert(resultant_primes.end(), v.begin(), v.end());
+    }
+}
+
+template<class Integer, class Container>
+void segmented_sieve(Integer lower_bound, Integer upper_bound, Container &resultant_primes)
+{
+    using boost::math::detail::linear_sieve_limit;
+    
+    Integer limit{static_cast<Integer>(std::floor(std::sqrt(static_cast<double>(upper_bound)))) + 1};
+    std::vector<Integer> primes {};
+    primes.reserve(static_cast<double>(limit) / std::log(static_cast<double>(limit)));
+
+    // Prepare for max value so you do not have to calculate this again
+    if(limit < linear_sieve_limit<Integer>)
+    {
+        boost::math::detail::linear_sieve(static_cast<Integer>(limit), primes);
+    }
+
+    else
+    {
+        boost::math::detail::linear_sieve(linear_sieve_limit<Integer>, primes);
+        boost::math::detail::segmented_sieve(linear_sieve_limit<Integer>, limit, primes, primes);
+    }
+
+    boost::math::detail::segmented_sieve(lower_bound, upper_bound, primes, resultant_primes);
+}
+
+template<class Integer, class Container>
+void sequential_segmented_sieve(Integer lower_bound, Integer upper_bound, Container &resultant_primes)
+{
+    const Integer L1_SIZE {32768};
+    const Integer interval {L1_SIZE * 8};
+    Integer current_lower_bound{lower_bound};
+    Integer current_upper_bound{current_lower_bound + interval};
+
+    if(current_upper_bound > upper_bound)
+    {
+        current_upper_bound = upper_bound;
+    }
+
+    std::size_t ranges {static_cast<std::size_t>((upper_bound - lower_bound) / interval)};
+
+    boost::math::detail::IntervalSieve sieve(current_lower_bound, current_upper_bound, resultant_primes, resultant_primes);
+    if(ranges == 0)
+    {
+        return;
+    }
+
+    for(std::size_t i {}; i < ranges; ++i)
+    {
+        current_lower_bound = current_upper_bound;
+        current_upper_bound += interval;
+        if(current_upper_bound > upper_bound)
+        {
+            current_upper_bound = upper_bound;
+        }
+        sieve.NewRange(current_lower_bound, current_upper_bound, resultant_primes);
+    }
+}
+} // End namespace detail
+
+template<class Integer>
+constexpr void prime_reserve(Integer upper_bound, std::vector<Integer> &prime_container)
+{
+    prime_container.reserve(static_cast<double>(upper_bound) / std::log(static_cast<double>(upper_bound)));
```

Comment:
I recommend that you just lift this equation (x / ln(x)) out, give it a name and a brief explanation of what it means.

Reply:
This one is, if I'm not mistaken, an underestimate of the true number of primes. (Not saying that _under_ estimate is bad, it's just a good name to differentiate from an overestimate.)